docs/modules/sql/pages/user-defined-types.adoc

```diff
@@ -1,173 +1,161 @@
 However, if the type hierarchy was not used in a mapping, any type in that hierarchy can be safely
 changed, and these changes will appear in the new mapping. This is because the links
-between types are symbolic (based on the name only), and they're only "materialized" once used in a mapping.
+between types are symbolic (based on the name only), and they are materialized only when used in a mapping.
 
-=== CREATE TYPE Examples
+=== Examples
+The following classes are used as a reference in the sections below to create types and mappings:
 
-NOTE: `EXTERNAL NAME` aliases are not supported for types, column names have to have exact
-same name as their corresponding Java/Portable/Compact class fields.
+[source,java]
+----
+package com.example;
 
-Java Type with auto-resolution for columns:
+class User implements Serializable {
+    public Long id;
+    public String name;
+    public Organization organization;
+}
 
-[source,sql]
-----
-CREATE TYPE MyType OPTIONS (
-    'format'='java',
-    'javaClass'='com.example.MyJavaClass'
-)
+class Organization implements Serializable {
+    public Long id;
+    public String name;
+    public Office office;
+}
+
+class Office implements Serializable {
+    public Long id;
+    public String name;
+}
 ----
 
-Java type with explicit columns:
+NOTE: The name of a type can differ from the one specified in the Java class or Portable/Compact/Avro schema. However, types must have distinct names within the set of names across all mappings and views as they share the same namespace.
 
+[#organization-office-types]
 [source,sql]
 ----
-CREATE TYPE MyType (
-    id BIGINT,
+CREATE TYPE Organization (
+    id BIGINT
     name VARCHAR,
-    other MyOtherType
+    office Office
 ) OPTIONS (
-    'format'='java',
-    'javaClass'='com.example.MyJavaClass'
-)
+    'javaClass'='com.example.Organization'
+);
+
+CREATE TYPE Office (
+    id BIGINT
+    name VARCHAR
+) OPTIONS (
+    'javaClass'='com.example.Office'
+);
 ----
 
-Portable Type:
+== Creating Mappings
+
+NOTE: The `organization` column is explicitly specified as `Organization` to prevent it being auto-resolved as a generic `OBJECT`, and therefore unable to query its sub-columns.
 
+[#users-mapping]
 [source,sql]
 ----
-CREATE TYPE MyPortableType (
+CREATE MAPPING users (
+    __key BIGINT,
     id BIGINT,
-    name VARCHAR
-) OPTIONS (
-    'format'='java',
-    'portableFactoryId'='1',
-    'portableClassId'='1'
-    -- 'portableVersion'='0' - specified by default
-)
+    name VARCHAR,
+    organization Organization
+) TYPE IMap OPTIONS (
+    'keyFormat'='bigint',
+    'valueFormat'='java',
+    'valueJavaClass'='com.example.User'
+);
 ----
 
-Compact Type:
+== Support for Cycles
+Cycles between types are only supported for Java format. However, this is limited to querying, not inserting or updating.
 
-[source,sql]
+NOTE: The existence of referenced types are only verified when the type is used in a `CREATE MAPPING` statement.
```

Comment:
```suggestion
NOTE: It is possible to reference a not yet existing type (including the type being defined) in type definition which is necessary to create cyclic structure. The existence of referenced types is only verified when the type is used in a `CREATE MAPPING` statement.
```
Without the addition I did not understand how this note relates to cyclic types.

"including the type being defined" - is that true?

Reply:
Yes, it is true. (fef8c8c2)